### PR TITLE
Update "modern" data model to match designs

### PR DESCRIPTION
**Pull request by ushkarev**
• both prisoners have a "role" in the non-association; options are pre-defined in an enum
• there is 1 "reason" for the non-association; options are pre-defined in an enum
• there is 1 "restriction type" for the non-association; options are pre-defined in an enum
• drop "incident report number" because it's not used and actually referred to Intelligence Reports

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/dto/SyncRequest.kt [hunk 1]
```diff
@@ -50,16 +50,17 @@ data class CreateSyncRequest(
...
-      firstPrisonerReason = firstPrisonerReason,
+      firstPrisonerRole = firstPrisonerReason.toRole(),
       secondPrisonerNumber = secondPrisonerNumber,
-      secondPrisonerReason = secondPrisonerReason,
-      restrictionType = restrictionType,
-      comment = comment ?: "",
+      secondPrisonerRole = secondPrisonerReason.toRole(),
+      reason = Reason.OTHER, // TODO: what's the "default" reason?
+      restrictionType = restrictionType.toRestrictionType(),
+      comment = comment ?: "No comment provided", // TODO: can we have a better message?
```
Comment (ushkarev) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/dto/SyncRequest.kt:58]: NB: previously the comment synced to `""` but this is not allowed by DB constraint

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/service/NonAssociationsMergeService.kt [hunk 2]
```diff
@@ -6,7 +6,6 @@ import org.slf4j.LoggerFactory
 import org.springframework.stereotype.Service
 import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.jpa.NonAssociation
 import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.jpa.repository.NonAssociationsRepository
-import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.jpa.NonAssociation as NonAssociationJPA
```
Comment (ushkarev) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/service/NonAssociationsMergeService.kt:9]: already imported directly as `NonAssociation `

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/service/SyncAndMigrateService.kt [hunk 3]
```diff
@@ -35,16 +35,14 @@ class SyncAndMigrateService(
...
-      restrictionType = updateSyncRequest.restrictionType
-      firstPrisonerReason = updateSyncRequest.firstPrisonerReason
-      secondPrisonerReason = updateSyncRequest.secondPrisonerReason
-      comment = updateSyncRequest.comment ?: ""
+    val na = with(nonAssociation) {
+      restrictionType = updateSyncRequest.restrictionType.toRestrictionType()
+      firstPrisonerRole = updateSyncRequest.firstPrisonerReason.toRole()
+      secondPrisonerRole = updateSyncRequest.secondPrisonerReason.toRole()
+      // TODO: can we have a better fall back message?
+      comment = updateSyncRequest.comment ?: "No comment provided"
```
Comment (ushkarev) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/service/SyncAndMigrateService.kt:45]: NB: an empty comment no longer overwrites to `""`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/dto/NonAssociation.kt [hunk 4]
```diff
@@ -13,15 +13,17 @@ data class NonAssociation(
...
+  val firstPrisonerRole: Role,
   @Schema(description = "Prisoner number to not associate", required = true, example = "D5678EF")
   val secondPrisonerNumber: String,
-  @Schema(description = "Reason why this prisoner should be kept apart from the other", required = true, example = "PERPETRATOR")
-  val secondPrisonerReason: NonAssociationReason,
+  @Schema(description = "Other prisoner’s role in the non-association", required = true, example = "PERPETRATOR")
+  val secondPrisonerRole: Role,
 
+  @Schema(description = "Reason why these prisoners should be kept apart", required = true, example = "BULLYING")
+  val reason: Reason,
```
Comment (ushkarev) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/dto/NonAssociation.kt:24]: The "reason" for a non-association applies to the whole thing, not either person involved.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/dto/NonAssociation.kt [hunk 5]
```diff
@@ -13,15 +13,17 @@ data class NonAssociation(
...
-  @Schema(description = "Reason why this prisoner should be kept apart from the other", required = true, example = "VICTIM")
-  val firstPrisonerReason: NonAssociationReason,
+  @Schema(description = "This prisoner’s role in the non-association", required = true, example = "VICTIM")
+  val firstPrisonerRole: Role,
   @Schema(description = "Prisoner number to not associate", required = true, example = "D5678EF")
   val secondPrisonerNumber: String,
-  @Schema(description = "Reason why this prisoner should be kept apart from the other", required = true, example = "PERPETRATOR")
-  val secondPrisonerReason: NonAssociationReason,
+  @Schema(description = "Other prisoner’s role in the non-association", required = true, example = "PERPETRATOR")
+  val secondPrisonerRole: Role,
```
Comment (ushkarev) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/dto/NonAssociation.kt:21]: Both people have a "role" that resulted in a non-association, not a "reason"

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/service/SyncAndMigrateService.kt [hunk 6]
```diff
@@ -53,9 +51,11 @@ class SyncAndMigrateService(
         closedBy = null
       } else {
         if (closedReason == null) {
+          // TODO: can we have a better message?
           closedReason = "UNDEFINED"
         }
         if (closedBy == null) {
+          // TODO: perhaps system user would be more appropriate here
```
Comment (Mjwillis) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/service/SyncAndMigrateService.kt:58]: Agree!

**Review by Mjwillis — APPROVED** [review]
LGTM

**Pull request merged**